src/sinks/splunk_hec/common/service.rs

```diff
@@ -38,19 +96,69 @@ impl Service<HecRequest> for HecService {
     type Error = crate::Error;
     type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
 
-    fn poll_ready(&mut self, _cx: &mut Context) -> std::task::Poll<Result<(), Self::Error>> {
-        Poll::Ready(Ok(()))
+    fn poll_ready(&mut self, cx: &mut Context) -> std::task::Poll<Result<(), Self::Error>> {
+        // Ready if indexer acknowledgements is disabled or there is room for
+        // additional pending acks. Otherwise, wait until there is room.
+        if self.ack_finalizer_tx.is_none() {
+            Poll::Ready(Ok(()))
+        } else {
+            match self.ack_slots.poll_acquire(cx) {
+                Poll::Ready(Some(permit)) => {
+                    if self.current_ack_slot.replace(permit).is_some() {
+                        Poll::Ready(Err("poll_ready called after a successful call".into()))
+                    } else {
+                        Poll::Ready(Ok(()))
+                    }
+                }
+                Poll::Ready(None) => Poll::Ready(Err(
+                    "Indexer acknowledgements semaphore unexpectedly closed".into(),
+                )),
+                Poll::Pending => Poll::Pending,
+            }
```

Comment:
There's a small bug here, I believe.

Since we first poll the semaphore for an available permit, we have to wait for the semaphore to give us a permit back before we actually check if we already have a permit.  If we assume that some downstream usage of our service is broken, and unintentionally calls `poll_ready` more than once, we _may_ deadlock due to the downstream caller waiting for us to tell them we're ready, while we wait for the semaphore to give us a permit that it may never be able to give us... because we're holding the last one.

It's slightly contrived, since the number for maximum pending acks is likely to be very big, but it's _technically_ possible.

I would probably rework the code so that we first match against `self.current_ack_slot.take()`, and if it's `Some(..)`, return the error, and otherwise, run through the logic you have after calling `poll_acquire`.  With that structure, you can also wrap the call to `poll_acquire` with `ready!` which will handle the `Poll::Pending` case for you.

Reply:
Ah I see, that makes sense. Will fix this. Just curious, re-referencing my question [here](https://github.com/vectordotdev/vector/pull/10135#discussion_r761567108), why are clients not supposed to call `poll_ready` after `Poll::Ready(Ok)`? I don't know why they would want to, but, in this case, I imagine we could check  if a permit already exists via `self.current_ack_slot.is_some()` and immediately return `Poll::Ready(Ok)` instead of `Err`.

Reply:
Yep, I'm fine with that. 👍🏻  (re: my response on the other question)